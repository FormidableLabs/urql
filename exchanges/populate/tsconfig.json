{
  "extends": "../../tsconfig.json",
  "include": ["src", "../shared/src"],
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "urql": ["../../node_modules/urql/src"],
      "*-urql": ["../../node_modules/*-urql/src"],
      "@urql/*": ["../../node_modules/@urql/*/src"],
      "shared": ["../../node_modules/shared/src"]
    }
  }
}
