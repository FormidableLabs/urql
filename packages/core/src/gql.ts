/* eslint-disable prefer-rest-params */
import { TypedDocumentNode } from '@graphql-typed-document-node/core';

import {
  DocumentNode,
  DefinitionNode,
  FragmentDefinitionNode,
  Kind,
} from 'graphql';

import { keyDocument, stringifyDocument } from './utils';

const applyDefinitions = (
  fragmentNames: Map<string, string>,
  target: DefinitionNode[],
  source: Array<DefinitionNode> | ReadonlyArray<DefinitionNode>
) => {
  for (let i = 0; i < source.length; i++) {
    const definition = source[i];
    if (source[i].kind === Kind.FRAGMENT_DEFINITION) {
      const name = (source[i] as FragmentDefinitionNode).name.value;
      const value = stringifyDocument(definition);
      // Fragments will be deduplicated according to this Map
      const prevValue = fragmentNames.get(name);
      if (prevValue === undefined) {
        fragmentNames.set(name, value);
        target.push(source[i]);
      } else if (process.env.NODE_ENV !== 'production' && prevValue !== value) {
        // Fragments with the same names is expected to have the same contents
        console.warn(
          '[WARNING: Duplicate Fragment] A fragment with name `' +
            name +
            '` already exists in this document.\n' +
            'While fragment names may not be unique across your source, each name must be unique per document.'
        );
      }
    } else {
      target.push(source[i]);
    }
  }
};

function gql<Data = any, Variables = object>(
  strings: TemplateStringsArray,
  ...interpolations: Array<TypedDocumentNode | DocumentNode | string>
): TypedDocumentNode<Data, Variables>;

function gql<Data = any, Variables = object>(
  string: string
): TypedDocumentNode<Data, Variables>;

function gql(/* arguments */) {
  const fragmentNames = new Map<string, string>();
  const definitions: DefinitionNode[] = [];
  const interpolations: DocumentNode[] = [];

  // Apply the entire tagged template body's definitions
  let body: string = Array.isArray(arguments[0])
    ? arguments[0][0]
    : arguments[0] || '';
  for (let i = 1; i < arguments.length; i++) {
    const value = arguments[i];
    if (value && value.kind === Kind.DOCUMENT) {
      interpolations.push(value);
    } else {
      body += value;
    }

    body += arguments[0][i];
  }

  // Apply the tag's body definitions
  applyDefinitions(fragmentNames, definitions, keyDocument(body).definitions);

  // Copy over each interpolated document's definitions
  for (let i = 0; i < interpolations.length; i++)
    applyDefinitions(fragmentNames, definitions, interpolations[i].definitions);

  return keyDocument({
    kind: Kind.DOCUMENT,
    definitions,
  });
}

export { gql };
