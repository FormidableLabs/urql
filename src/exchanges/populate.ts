import { pipe, tap, map } from 'wonka';
import { Exchange, Operation } from '../types';
import {
  DocumentNode,
  buildClientSchema,
  visitWithTypeInfo,
  TypeInfo,
  FragmentDefinitionNode,
  SelectionSetNode,
  GraphQLSchema,
  IntrospectionQuery,
} from 'graphql';
import { visit } from 'graphql';

interface ExchangeArgs {
  schema: IntrospectionQuery;
}

/** An exchange for auto-populating mutations with a required response body. */
export const populateExchange = ({
  schema: ogSchema,
}: ExchangeArgs): Exchange => ({ forward }) => {
  const schema = buildClientSchema(ogSchema);
  /** List of operation keys that have already been parsed. */
  let parsedOperations: Record<string, boolean | undefined> = {};
  /** List of operation keys that have not been torn down. */
  let activeOperations: Record<string, boolean | undefined> = {};
  /** Collection of fragments used by the user. */
  let fragments: UserFragmentMap = {};
  /** Collection of type fragments. */
  let selections: TypeFragmentMap = {};

  /** Handle mutation and inject selections + fragments. */
  const handleIncomingMutation = (op: Operation) => {
    if (op.operationName !== 'mutation') {
      return op;
    }

    const activeSelections = Object.entries(selections).reduce(
      (state, [key, value]) => ({
        ...state,
        [key]: value.filter(s => activeOperations[s.key]),
      }),
      selections
    );

    return {
      ...op,
      query: addFragmentsToQuery({
        schema,
        selections: activeSelections,
        fragments,
        query: op.query,
      }),
    };
  };

  /** Handle query and extract fragments. */
  const handleIncomingQuery = ({ key, operationName, query }: Operation) => {
    activeOperations = addKey(activeOperations, key);

    if (operationName !== 'query' || parsedOperations[key]) {
      return;
    }

    parsedOperations = addKey(parsedOperations, key);

    const {
      fragments: newFragments,
      selections: newSelections,
    } = makeFragmentsFromQuery({
      schema,
      query,
    });

    fragments = newFragments.reduce(
      (state, fragment) => ({
        ...state,
        [fragment.name.value]: fragment,
      }),
      fragments
    );

    selections = newSelections.reduce((state, { selections, type }) => {
      const current = state[type] || [];
      const entry: TypeFragment = {
        key,
        fragment: {
          kind: 'FragmentDefinition',
          typeCondition: {
            kind: 'NamedType',
            name: {
              kind: 'Name',
              value: type,
            },
          },
          name: {
            kind: 'Name',
            value: `${type}_PopulateFragment_${current.length}`,
          },
          selectionSet: selections,
        },
        type,
      };
      return {
        ...state,
        [type]: [...current, entry],
      };
    }, selections);
  };

  const handleIncomingTeardown = ({ key, operationName }: Operation) => {
    if (operationName !== 'teardown') {
      return;
    }

    activeOperations = removeKey(activeOperations, key);
  };

  return ops$ => {
    return pipe(
      ops$,
      tap(handleIncomingTeardown),
      tap(handleIncomingQuery),
      map(handleIncomingMutation),
      forward
    );
  };
};

type UserFragmentMap<T extends string = string> = Record<
  T,
  FragmentDefinitionNode
>;

type TypeFragmentMap<T extends string = string> = Record<T, TypeFragment[]>;

interface TypeFragment {
  /** Operation key where selection set is being used. */
  key: number;
  /** Selection set. */
  fragment: FragmentDefinitionNode;
  /** Type of selection. */
  type: string;
}

interface MakeFragmentsFromQueryArg {
  schema: GraphQLSchema;
  query: DocumentNode;
}

/** Creates fragments object from query */
export const makeFragmentsFromQuery = ({
  schema,
  query,
}: MakeFragmentsFromQueryArg) => {
  let selections: { selections: SelectionSetNode; type: string }[] = [];
  let fragments: FragmentDefinitionNode[] = [];
  const typeInfo = new TypeInfo(schema);

  visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: node => {
        if (!node.selectionSet) {
          return undefined;
        }

        // @ts-ignore
        const t = typeInfo.getType().ofType;

        if (!t) {
          return undefined;
        }

        selections = [
          ...selections,
          { selections: node.selectionSet, type: t },
        ];
      },
      FragmentDefinition: node => {
        fragments = [...fragments, node];
      },
    })
  );

  return { selections, fragments };
};

interface AddFragmentsToQuery {
  schema: GraphQLSchema;
  query: DocumentNode;
  selections: Record<string, Omit<TypeFragment, 'key'>[]>;
  fragments: UserFragmentMap;
}

export const addFragmentsToQuery = ({
  schema,
  query,
  selections,
  fragments,
}: AddFragmentsToQuery) => {
  const typeInfo = new TypeInfo(schema);
  let additionalFragments: Record<string, FragmentDefinitionNode> = {};

  return visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: {
        enter: node => {
          const directive =
            node.directives &&
            node.directives.find(d => d.name.value === 'populate');

          if (!directive) {
            return;
          }

          const t = typeInfo.getType() as any;
          const type = t.ofType || t.toString();

          const existingSelections =
            (node.selectionSet && node.selectionSet.selections) || [];
          const newSelections = selections[type]
            ? selections[type].map(({ fragment }) => {
                // Add fragment for Document node
                additionalFragments = {
                  ...additionalFragments,
                  [fragment.name.value]: fragment,
                };

                return {
                  kind: 'FragmentSpread',
                  name: {
                    kind: 'Name',
                    value: fragment.name.value,
                  },
                };
              })
            : [];

          return {
            ...node,
            directives: (node.directives as any).filter(
              d => d.name.value !== 'populate'
            ),
            selectionSet: {
              kind: 'SelectionSet',
              selections: [...newSelections, ...existingSelections],
            },
          };
        },
      },
      Document: {
        leave: node => {
          return {
            ...node,
            definitions: [
              ...node.definitions,
              ...Object.values(additionalFragments),
              ...Object.values(fragments),
            ],
          };
        },
      },
    })
  );
};

const addKey = (
  s: Record<string, boolean | undefined>,
  key: number | string
) => ({
  ...s,
  [key]: true,
});

const removeKey = (
  s: Record<string, boolean | undefined>,
  key: number | string
) => ({
  ...s,
  [key]: false,
});
