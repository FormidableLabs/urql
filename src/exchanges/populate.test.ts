import { extractSelectionsFromQuery, addFragmentsToQuery } from './populate';
import {
  buildSchema,
  print,
  FragmentDefinitionNode,
  visit,
  SelectionSetNode,
} from 'graphql';
import gql from 'graphql-tag';

const schemaDef = `
  type User {
    id: ID!
    name: String!
    age: Int!
  }

  type Todo {
    id: ID!
    text: String!
    creator: User!
  }

  type Query {
    todos: [Todo]
  }

  type Mutation {
    addTodo: [Todo]
  }
`;

const query = gql`
  query {
    todos {
      id
      name
      creator {
        id
      }
    }
  }

  fragment TodoFragment on Todo {
    id
  }
`;

const schema = buildSchema(schemaDef);

describe('makeFragmentsFromQuery', () => {
  const arg = {
    schema,
    query,
  };

  describe('new selections', () => {
    it('are created', () => {
      const r = extractSelectionsFromQuery(arg);
      expect(r.selections.length).toBe(2);
      expect(r).toMatchSnapshot();
    });
  });

  describe('user declared fragments', () => {
    it('are extracted', () => {
      const r = extractSelectionsFromQuery(arg);
      expect(
        r.fragments.filter(f => f.name.value === 'TodoFragment')
      ).toHaveLength(1);
    });
  });
});

describe('addFragmentsToQuery', () => {
  const mutation = gql`
    mutation MyMutation {
      addTodo @populate
    }
  `;

  const arg = {
    schema,
    query: mutation,
    typeFragments: {
      Todo: [
        {
          key: 1234,
          fragment: gql`
            fragment Type_PopulateFragment_0 on Todo {
              id
            }
          `.definitions[0],
          type: 'Todo',
        },
      ],
    },
    userFragments: {
      MyFragment: gql`
        fragment MyFragment on User {
          id
          name
        }
      `.definitions[0],
    },
  };

  it('returns result matching snapshot', () => {
    const q = addFragmentsToQuery(arg);
    expect(print(q)).toMatchInlineSnapshot(`
      "mutation MyMutation {
        addTodo {
          ...Type_PopulateFragment_0
        }
      }

      fragment Type_PopulateFragment_0 on Todo {
        id
      }

      fragment MyFragment on User {
        id
        name
      }
      "
    `);
  });

  describe('user provided fragments', () => {
    it('are added to query', () => {
      const r = addFragmentsToQuery(arg);

      let fragmentDefs: FragmentDefinitionNode[] = [];
      visit(r, {
        FragmentDefinition: node => {
          fragmentDefs = [...fragmentDefs, node];
        },
      });

      expect(
        fragmentDefs.filter(d => d.name.value === 'MyFragment')
      ).toHaveLength(1);
    });
  });

  describe('new type fragments', () => {
    it('are added to query', () => {
      const r = addFragmentsToQuery(arg);

      let addedFragments: FragmentDefinitionNode[] = [];
      visit(r, {
        FragmentDefinition: node => {
          addedFragments = [...addedFragments, node];
        },
      });

      expect(
        addedFragments.filter(s => s === arg.typeFragments.Todo[0].fragment)
      ).toHaveLength(1);
    });

    describe('on empty', () => {
      it('typename is returned', () => {
        const r = addFragmentsToQuery({ ...arg, typeFragments: {} });
        expect(print(r)).toMatchInlineSnapshot(`
          "mutation MyMutation {
            addTodo {
              __typename
            }
          }

          fragment MyFragment on User {
            id
            name
          }
          "
        `);
      });
    });
  });
});
