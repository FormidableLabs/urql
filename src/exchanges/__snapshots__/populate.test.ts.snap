// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeFragmentsFromQuery new selections are created 1`] = `
Object {
  "fragments": Array [
    Object {
      "directives": Array [],
      "kind": "FragmentDefinition",
      "name": Object {
        "kind": "Name",
        "value": "TodoFragment",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "Todo",
        },
      },
    },
  ],
  "selections": Array [
    Object {
      "selections": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
            "selectionSet": undefined,
          },
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "name",
            },
            "selectionSet": undefined,
          },
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "creator",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
          },
        ],
      },
      "type": "Todo",
    },
    Object {
      "selections": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
            "selectionSet": undefined,
          },
        ],
      },
      "type": "User",
    },
  ],
}
`;
